library/src/com/google/maps/android/clustering/algo/VisibleNonHierarchicalDistanceBasedAlgorithm.java

```diff
@@ -0,0 +1,224 @@
+ * 3. Add all items that are within a certain distance to the cluster. <br>
+ * 4. Move any items out of an existing cluster if they are closer to another cluster. <br>
+ * 5. Remove those items from the list of candidate clusters.
+ * <p/>
+ * Clusters have the center of the first element (not the centroid of the items within it).
+ */
+public class VisibleNonHierarchicalDistanceBasedAlgorithm<T extends ClusterItem>
+        implements Algorithm<T>, GoogleMap.OnCameraChangeListener {
+
+    public static final int MAX_DISTANCE_AT_ZOOM = 100; // essentially 100 dp.
+
+    /**
+     * Any modifications should be synchronized on mQuadTree.
+     */
+    private final Collection<QuadItem<T>> mItems = new ArrayList<QuadItem<T>>();
+
+    /**
+     * Any modifications should be synchronized on mQuadTree.
+     */
+    private final PointQuadTree<QuadItem<T>> mQuadTree = new PointQuadTree<QuadItem<T>>(0, 1, 0, 1);
+
+    private static final SphericalMercatorProjection PROJECTION = new SphericalMercatorProjection(1);
+    private final int mScreenWidth;
+    private final int mScreenHeight;
+    private LatLng mMapCenter;
+
+    public VisibleNonHierarchicalDistanceBasedAlgorithm(int screenWidth, int screenHeight) {
+        mScreenWidth = screenWidth;
+        mScreenHeight = screenHeight;
+    }
+
+    @Override
+    public void addItem(T item) {
+        final QuadItem<T> quadItem = new QuadItem<T>(item);
+        synchronized (mQuadTree) {
+            mItems.add(quadItem);
+            mQuadTree.add(quadItem);
+        }
+    }
+
+    @Override
+    public void addItems(Collection<T> items) {
+        for (T item : items) {
+            addItem(item);
+        }
+    }
+
+    @Override
+    public void clearItems() {
+        synchronized (mQuadTree) {
+            mItems.clear();
+            mQuadTree.clear();
+        }
+    }
+
+    @Override
+    public void removeItem(T item) {
+        // TODO: delegate QuadItem#hashCode and QuadItem#equals to its item.
+        throw new UnsupportedOperationException("VisibleNonHierarchicalDistanceBasedAlgorithm.remove not implemented");
+    }
+
+    @Override
+    public Set<? extends Cluster<T>> getClusters(double zoom) {
+        final int discreteZoom = (int) zoom;
+
+        final double zoomSpecificSpan = MAX_DISTANCE_AT_ZOOM / Math.pow(2, discreteZoom) / 256;
+
+        final Set<QuadItem<T>> visitedCandidates = new HashSet<QuadItem<T>>();
+        final Set<Cluster<T>> results = new HashSet<Cluster<T>>();
+        final Map<QuadItem<T>, Double> distanceToCluster = new HashMap<QuadItem<T>, Double>();
+        final Map<QuadItem<T>, StaticCluster<T>> itemToCluster = new HashMap<QuadItem<T>, StaticCluster<T>>();
+
+        synchronized (mQuadTree) {
+
+            Bounds visibleBounds = getVisibleBounds(discreteZoom);
+
+            Collection<QuadItem<T>> items = mQuadTree.search(visibleBounds);
+
+            for (QuadItem<T> candidate : items) {
```

Comment:
As this block is straight from the NonHierarchicalDistanceBasedAlgorithm, it'd be great to pull it out into a separate method, on the parent class, when you subclass it.